### PR TITLE
Support some HttpClientAction settings and log debug message instead of an error

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspire/issues/4233

### CODE REVIEW

File: src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs [hunk 1]
```diff
@@ -190,4 +192,15 @@ public override void SetCompositeCredentials(object state, ChannelCredentials ch
         {
         }
     }
+
+    private static class Log
+    {
+        private static readonly Action<ILogger, string, Exception?> _unsupportedHttpClientActions =
+            LoggerMessage.Define<string>(LogLevel.Debug, new EventId(1, "UnsupportedHttpClientActions"), "The ConfigureHttpClient method is not supported when creating gRPC clients. Unable to apply configuration to client with the name '{ClientName}'.");
```
Comment (amcasey) [inline src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs:199]: This appears to indicate a misconfiguration - maybe it should be a warning?

File: src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs [hunk 2]
```diff
@@ -75,7 +77,7 @@ private CallInvoker CreateInvoker(EntryKey key)
             var httpClientFactoryOptions = _httpClientFactoryOptionsMonitor.Get(name);
             if (httpClientFactoryOptions.HttpClientActions.Count > 0)
```
Comment (amcasey) [inline src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs:78]: To confirm, the HttpClientActions will just be ignored?

**Review by amcasey — APPROVED** [review]
Feels like a Warning to me, but I leave it up to you.

File: src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs (hunk 1 above)
Comment (JamesNK) [reply src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs]: I've changed the log message content. I think debug is the right level for the new message.

File: src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs (hunk 2 above)
Comment (JamesNK) [reply src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs:87]: I've come up with a semi-hacky way to make them used. Actions now run on a dummy HttpClient and values are then partially applied to gRPC client. It's not perfect, but should cover most scenarios, and a log message provides more info.

File: src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs [hunk 3]
```diff
@@ -75,12 +77,58 @@ private CallInvoker CreateInvoker(EntryKey key)
...
+                foreach (var applyOptions in httpClientFactoryOptions.HttpClientActions)
+                {
+                    applyOptions(httpClient);
+                }
+
+                // Copy configuration from HttpClient to GrpcChannel/CallOptions.
+                // This configuration should be overriden by gRPC specific config methods.
+                if (clientFactoryOptions.Address == null)
+                {
+                    clientFactoryOptions.Address = httpClient.BaseAddress;
```
Comment (amcasey) [inline src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs:100]: Would it be interesting to log when the HttpClient value is present but ignored?

File: src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs [hunk 4]
```diff
@@ -75,12 +77,58 @@ private CallInvoker CreateInvoker(EntryKey key)
...
+                    // Don't replace or add new header values if the header name has already been set.
+                    clientFactoryOptions.CallOptionsActions.Add(callOptionsContext =>
+                    {
+                        var metadata = callOptionsContext.CallOptions.Headers ?? new Metadata();
+                        foreach (var entry in defaultHeaders)
+                        {
+                            // grpc requires header names are lower case before being added to collection.
+                            var resolvedKey = entry.Key.ToLower(CultureInfo.InvariantCulture);
+
+                            if (metadata.Get(resolvedKey) == null)
```
Comment (amcasey) [inline src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs:118]: Would it be interesting to log when it's not null and values form the HttpClient are ignored?

**Review by amcasey — APPROVED** [review]
LGTM

File: src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs (hunk 3 above)
Comment (JamesNK) [reply src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs:100]: I don't think we need to go to this detail.

File: src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs (hunk 4 above)
Comment (JamesNK) [reply src/Grpc.Net.ClientFactory/Internal/GrpcCallInvokerFactory.cs:118]: I don't think we need to go to this detail.

**Pull request merged**